src/docker-images/job-insighter/insight.py

```diff
@@ -0,0 +1,562 @@
+            "num": 0,
+        })
+        url = urllib.parse.urljoin(restful_url, "/ListJobsV2") + "?" + params
+        obj = request_with_error_handling(url)
+
+        jobs = walk_json_field_safe(obj, "runningJobs")
+        running_job_ids = [job.get("jobId") for job in jobs]
+    except:
+        logger.exception("Failed to get running job ids for vc %s.", vc_name)
+
+    return running_job_ids
+
+
+def get_running_job_ids(restful_url, vc_info):
+    running_job_ids = []
+    for vc in vc_info:
+        vc_running_job_ids = \
+            get_vc_running_job_ids(restful_url, vc.get("vcName"))
+        running_job_ids.extend(vc_running_job_ids)
+    return running_job_ids
+
+
+def set_job_insight(restful_url, job_id, insight):
+    params = urllib.parse.urlencode({
+        "jobId": job_id,
+        "userName": "Administrator",
+    })
+    url = urllib.parse.urljoin(restful_url, "/JobInsight") + "?" + params
+    resp = requests.post(url, data=json.dumps(insight))
+    return resp
+
+
+def avg(ts):
+    """Calculates the average of the timeseries.
+
+    Args:
+        ts: A timeseries list of [time, value].
+
+    Returns:
+        Average of the timeseries.
+    """
+    return sum([float(v[1]) for v in ts]) / len(ts)
+
+
+def timespan(ts):
+    """Calculates the time span of the timeseries.
+
+    Args:
+        ts: A timeseries list of [time, value].
+
+    Returns:
+        The time span of the timeseries.
+    """
+    return ts[-1][0] - ts[0][0]
+
+
+class JobUtil(object):
+    def __init__(self, job_id, utils, since, end):
+        self.job_id = job_id
+        self.since = since
+        self.end = end
+
+        # Calculate average usage
+        self.gpu = {
+            uuid: avg(ts) for uuid, ts in utils["gpu"].items()
+        }
+        self.gpu_memory = {
+            uuid: avg(ts) for uuid, ts in utils["gpu_memory"].items()
+        }
+        self.cpu = {
+            pod_name: avg(ts) / 100 for pod_name, ts in utils["cpu"].items()
+        }
+        self.memory = {
+            pod_name: avg(ts) for pod_name, ts in utils["memory"].items()
+        }
+
+        # Compute time span
+        timespans = [timespan(ts) for _, ts in utils["gpu"].items()]
+        timespans.extend(
+            [timespan(ts) for _, ts in utils["gpu_memory"].items()])
+        timespans.extend([timespan(ts) for _, ts in utils["cpu"].items()])
+        timespans.extend([timespan(ts) for _, ts in utils["memory"].items()])
+        self.timespan = min(timespans)
+
+    def __repr__(self):
+        return str(self.__dict__)
+
+
+class Insight(object):
+    def __init__(self, job_util, node_spec):
+        self.job_id = job_util.job_id
+        self.since = job_util.since
+        self.end = job_util.end
+
+        # Resource usage pattern from job_util
+        self.job_timespan = job_util.timespan
+        self.num_gpus = len(job_util.gpu)
+        self.idle_gpus = sorted([u for u, v in job_util.gpu.items() if v == 0])

```

Comment:
Please do not put computation or function call inside `__init__`. This is not conventional.